xtm/xtm_api.cc

```diff
@@ -0,0 +1,364 @@
+xtm_queue_delete(struct xtm_queue *queue)
+{
+	int rc = 0;
+	if (((queue->flags & XTM_QUEUE_MUST_CLOSE_READFD) != 0) &&
+	    close(queue->producer_read_fd) < 0)
+		rc = -1;
+	if (queue->producer_read_fd != queue->producer_write_fd &&
+	    close(queue->producer_write_fd) < 0)
+		rc = -1;
+	if (((queue->flags & XTM_QUEUE_MUST_CLOSE_READFD) != 0) &&
+	    close(queue->consumer_read_fd) < 0)
+		rc = -1;
+	if (queue->consumer_read_fd != queue->consumer_write_fd &&
+	    close(queue->consumer_write_fd) < 0)
+		rc = -1;
+	free(queue);
+	return rc;
+}
+
+int
+xtm_queue_notify_consumer(struct xtm_queue *queue)
+{
+	return notify_fd(queue->consumer_write_fd);
+}
+
+int
+xtm_queue_notify_producer(struct xtm_queue *queue)
+{
+	return notify_fd(queue->producer_write_fd);
+}
+
+int
+xtm_queue_probe(struct xtm_queue *queue)
+{
+	if (queue->queue.free_count() == 0) {
+		errno = ENOBUFS;
+		return -1;
+	}
+	return 0;
+}
+
+unsigned
+xtm_queue_count(struct xtm_queue *queue)
+{
+	return queue->queue.count();
+}
+
+int
+xtm_queue_push_fun(struct xtm_queue *queue, xtm_queue_fun_t fun,
+		   void *fun_arg, unsigned flags)
+{
+	union xtm_msg xtm_msg;
+	xtm_msg.fun = fun;
+	xtm_msg.fun_arg = fun_arg;
+	if (queue->queue.put(&xtm_msg, 1) != 0)
+		return 0;
+
+	assert((flags & (~XTM_QUEUE_PRODUCER_NEEDS_NOTIFICATIONS)) == 0);
+	if ((flags & XTM_QUEUE_PRODUCER_NEEDS_NOTIFICATIONS) == 0)
+		goto error;
+
+	__atomic_store_n(&queue->is_producer_should_be_notified, true,
+			 __ATOMIC_SEQ_CST);
+
+	/*
+	 * It is necessary to explain why we are trying to push and then set
+	 * is_producer_should_be_notified flag and then tries to push again!
+	 * The fact is that consumer thread could invoke funs and checked
+	 * is_producer_should_be_notified, BEFORE it was set in this function.
+	 * In this case we try to push again (consumer thread freed space in
+	 * queue in this case).
+	 */
+	if (queue->queue.put(&xtm_msg, 1) != 0)
+		return 0;
+
+error:
+	errno = ENOBUFS;
+	return -1;
+}
+
+int
+xtm_queue_consumer_fd(struct xtm_queue *queue)
+{
+	return queue->consumer_read_fd;
+}
+
+int
+xtm_queue_producer_fd(struct xtm_queue *queue)
+{
+	return queue->producer_read_fd;
+}
+
+unsigned
+xtm_queue_invoke_funs_all(struct xtm_queue *queue)
+{
+	struct xtm_scsp_queue_read_iterator<xtm_msg> iter;
+	const union xtm_msg *xtm_msg;
+	unsigned cnt = 0;
+
+	iter.begin(&queue->queue);
+	while((xtm_msg = iter.read()) != nullptr) {
+		xtm_msg->fun(xtm_msg->fun_arg);
+		cnt++;
+	}
+	iter.end();
+	return cnt;
+}
+
+int
+xtm_queue_push_ptr(struct xtm_queue *queue, void *ptr, unsigned flags)
+{
+	union xtm_msg xtm_msg;
+	xtm_msg.ptr = ptr;
+	if (queue->queue.put(&xtm_msg, 1) != 0)
+		return 0;
+
+	assert((flags & (~XTM_QUEUE_PRODUCER_NEEDS_NOTIFICATIONS)) == 0);
+	if ((flags & XTM_QUEUE_PRODUCER_NEEDS_NOTIFICATIONS) == 0)
+		goto error;
+
+	__atomic_store_n(&queue->is_producer_should_be_notified, true,
+			 __ATOMIC_SEQ_CST);
+
+	/*
+	 * It is necessary to explain why we are trying to push and then set
```

Comment:
You should just reference the comment in xtm_queue_push_fun(). This way you would not get two different versions when someone would change one of them (e. g. for grammatics ;-))

Reply:
Fixed